From: Hendrik Brueckner <brueckner@redhat.com>
Date: Mon, 19 Jul 2010 11:47:54 -0400
Subject: [s390] qeth: don't allow layer switch with open interface
Message-id: <1279540075-25959-1-git-send-email-brueckner@redhat.com>
Patchwork-id: 26943
O-Subject: [RHEL5.6 PATCH 1/2] [s390x] qeth: do not allow layer switch with open
	network interface
Bugzilla: 612195
RH-Acked-by: Pete Zaitcev <zaitcev@redhat.com>

Description
-----------
It is rare possible the kernel crashes in case the layer2
attribute of a qeth device is changed during the network
interface is open.

qeth establish a manipulated ARP constructor for layer 3
devices. Changing from the manipulated constructor to the
default system constructor and vice versa causes problems
in the Linux neighbour code.

We do not allow switching the qeth layer2 attribute as
long as the network interface is open.

Bugzilla
--------
BZ 612195
https://bugzilla.redhat.com/show_bug.cgi?id=612195

Upstream status of the patch
----------------------------
The problem is specific to the RHEL 5 release. The design of the
upstream qeth driver (rework with 2.6.26) prevents this kind of
error.

Test status
-----------
The patch has been tested and fixes the problem.
The fix has been verified by the IBM test department.

diff --git a/drivers/s390/net/qeth_sys.c b/drivers/s390/net/qeth_sys.c
index 4102f87..1579977 100644
--- a/drivers/s390/net/qeth_sys.c
+++ b/drivers/s390/net/qeth_sys.c
@@ -711,8 +711,7 @@ qeth_dev_layer2_store(struct device *dev, struct device_attribute *attr, const c
 	if (!card)
 		return -EINVAL;
 
-	if (((card->state != CARD_STATE_DOWN) &&
-	     (card->state != CARD_STATE_RECOVER)))
+	if (card->state != CARD_STATE_DOWN)
 		return -EPERM;
 
 	i = simple_strtoul(buf, &tmp, 16);
